api/pkg/handler/v1/provider/openstack.go

```diff
@@ -417,6 +417,83 @@ func getOpenstackSubnets(userInfo *provider.UserInfo, seedsGetter provider.Seeds
 	return apiSubnetIDs, nil
 }
 
+func OpenstackAvailabilityZoneEndpoint(seedsGetter provider.SeedsGetter, presetsProvider provider.PresetProvider, userInfoGetter provider.UserInfoGetter) endpoint.Endpoint {
+	return func(ctx context.Context, request interface{}) (interface{}, error) {
+		req, ok := request.(OpenstackReq)
+		if !ok {
+			return nil, fmt.Errorf("incorrect type of request, expected = OpenstackReq, got = %T", request)
+		}
+		userInfo, err := userInfoGetter(ctx, "")
+		if err != nil {
+			return nil, common.KubernetesErrorToHTTPError(err)
+		}
+
+		datacenterName := req.DatacenterName
+		_, datacenter, err := provider.DatacenterFromSeedMap(userInfo, seedsGetter, datacenterName)
+		if err != nil {
+			return nil, fmt.Errorf("error getting dc: %v", err)
+		}
+
+		username, password, domain, tenant, tenantID, err := getOpenstackCredentials(userInfo, req.Credential, req.Username, req.Password, req.Domain, req.Tenant, req.TenantID, presetsProvider)
```

Comment:
Nit: can we wrap these into a struct? they are passed around a lot so it would make sense.

Reply:
Sounds good to me, but we basically do this for all providers... Maybe it would make sense to do it in a seperate PR to "refactor" this for every provider? cc @zreigz WDYT?

Reply:
sounds good to me

Reply:
Created an issue for now, to definitely not forget it: https://github.com/kubermatic/kubermatic/issues/5557